executor/windowfunc/window_funcs.go

```diff
@@ -0,0 +1,49 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package windowfuncs
+
+import (
+	"unsafe"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// PartialResult represents data structure to store the partial result for the
+// aggregate functions. Here we use unsafe.Pointer to allow the partial result
+// to be any type.
+type PartialResult unsafe.Pointer
+
+// WindowFunc is the interface for processing window functions.
+type WindowFunc interface {
+	// ProcessOneChunk processes one chunk and write results to chunk.
+	ProcessOneChunk(sctx sessionctx.Context, rows []chunk.Row, pr PartialResult, dest *chunk.Chunk, remainedRows int) ([]chunk.Row, int, error)
```

Comment:
How about changing `remainedRows` to `numRowsToProcess` and maintaining this value in the caller side? Thus the return value can be simplified to `error` and we don't need to care about what does `remainedRows` mean.

That is:

```suggestion
	ProcessOneChunk(sctx sessionctx.Context, rows []chunk.Row, numRowsToProcess int, pr PartialResult, dest *chunk.Chunk) (error)
```

Reply:
- It is not `numRowsToProcess`, but `min(remainingRowsInChunk, remainingRowsInGroup)` to avoid processing more rows than this chunk or group needed.
- I think it is better to keep it here because it frees the caller to know how many rows will the callee processed and it is consistent with `ExhaustResult`.

Reply:
How about setting `numRowsToProcess` to `min(len(rows), remainingRowsInChunk, remainingRowsInGroup)`? Thus the function signature can be simplified, and the window function implementation don't need to consider the staffs that not related to the function itself.

The number of window function implementations can be very large, especially when we need to write a different implementations for different input types of the same window function. I think it's valuable to keep the interface simpler.

From the caller's perspective, I think this interface is more friendly to the case of multiple functions inside a WindowExec. For example, in the `noFrameWindowProcessor`:

```go
func (p *noFrameWindowProcessor) consumeGroupRows(ctx sessionctx.Context, rows []chunk.Row, chk *chunk.Chunk, remained int) ([]chunk.Row, int, error) {
	numRowsToProcess := min(len(rows), remained)
	for i := range p.WindowFuncs {
		err := p.WindowFuncs[i].ProcessOneChunk(ctx, rows, numRowsToProcess, p.partialResult[i], chk)
		// handle err
	}
	return rows[numRowsToProcess:], remained-numRowsToProcess, nil
}
```

From the window function's perspective, it only needs to handle numRowsToProcess input rows, only change the content of output parameters, these are `chk` and partial result, leave the input parameters unchanged, for example, in row_number():
```go
func (wf *rowNumber) ProcessOneChunk(sctx sessionctx.Context, rows []chunk.Row, numRowsToProcess int, pr PartialResult, dest *chunk.Chunk) error {
	p := (*partialResult4RowNumber)(pr)
	for i := 0; i < numRowsToProcess; i++ {
		p.curIdx++
		dest.AppendInt64(wf.ordinal, p.curIdx)
 	}
	return nil
}
```

Reply:
Not all functions perform like `row_number`, some functions also need to know rows before the current row, while some functions also need to know rows after the current row, like `lead` and `lag`. So it is difficult to know how many rows the function really proceed.

Reply:
These infos can be stored in the partial result struct like apache hive. For example `LeadBuffer`: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFLead.java#L81

Reply:
Then what about lag? The main problem here is you cannot know the number of rows processed without knowing the specific function.

Reply:
for `lag()`, there is also a `LagBuffer`: https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDAFLag.java#L85

Reply:
"The main problem here is you cannot know the number of rows processed without knowing the specific function.".

Reply:
The partial result is binded to the specific function, each function has its own partial result.

Reply:
I mean, in TiDB, `LeadBuffer` can be stored in the partial result of `Lead()`

Reply:
The question here is we cannot update the `numRowsToProcess` correctly without knowing the function. Take the previous code:
```go
func (p *noFrameWindowProcessor) consumeGroupRows(ctx sessionctx.Context, rows []chunk.Row, chk *chunk.Chunk, remained int) ([]chunk.Row, int, error) {
	numRowsToProcess := min(len(rows), remained)
	for i := range p.WindowFuncs {
		err := p.WindowFuncs[i].ProcessOneChunk(ctx, rows, numRowsToProcess, p.partialResult[i], chk)
		// handle err
	}
	return rows[numRowsToProcess:], remained-numRowsToProcess, nil
}
```
And assumes the function is `lead`, `remained` is 10, `len(rows)` is 5, and current rows is the first rows in the group, so `numRowsToProcess` is 5 here, but you can only process 4 rows for `lead` because you do not know row 6. So `numRowsToProcess` depends on each function.

Reply:
Updated, PTAL.